posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md

```diff
@@ -0,0 +1,209 @@
+
+- SciPy uses markdown templates for reporting bugs and feature requests through Issues. We can migrate them to use [GitHub Issue forms](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/syntax-for-issue-forms) in YAML for a better user experience.
+
+- SciPy CI build time takes around ~10 minutes with the Meson build system. It is an improvement over the existing `setuptools,` but for faster iterations, we need to capitalize on caching to reduce the total run-time of CI.
+
+- SciPy is currently tested on a variety of operating systems and architecture. We want to ensure the compatibility of the new build system for the same. A CI matrix should define the combination of variations on the build and the ideal target.
+
+- SciPy currently features benchmarks that are not automated on the CI. We want to automate this process on the CI itself by tagging Pull Requests with specific labels or triggering the same during releases.
+
+- Scipy currently supports X86, X64, PPC64, and ARM64 architecture through the [Azure DevOps](https://azure.microsoft.com/en-in/services/devops/) pipelines. We want to ensure that the compatibility remains the same with the new build system.
+
+- SciPy uses parallel builds across the Azure DevOps pipelines and macOS GitHub Actions along with the test scripts. We want to consider what parts of the entire matrix we wish to run on the CI providers we choose to balance the overhead of using multiple providers.
+
+- SciPy currently lacks any developer documentation for CI providers. Documenting the CI systems and the build matrix is necessary for developers to cross-reference them in future improvements. We can also document how to emulate CI runs locally through specific tools.
+
+## Implementing the local CI run
+
+The project's first step was to ensure that we could run the GitHub Actions CI locally to ensure faster iterations on the failures we encountered. GitHub does not feature an official tool to make this possible; hence we had to look at third-party alternatives like Act.
+
+`act` is a tool offered by Nektos which provides a handy way to run GitHub Actions locally. It gives a quick way to validate your changes on the CI locally, without committing/pushing your changes to the workflows to trigger and validate them. It leads to fast feedback and compatibility as a local task runner to validate all our CI jobs.
+
+You can try out `act` yourself by cloning [Ralf’s fork](https://github.com/rgommers/scipy) and referencing the official documentation for the [`act` usage](https://github.com/rgommers/scipy/blob/meson/ACT-USAGE.md). It keeps the GitHub Action notifications clutter-free and reduces the chance of accidentally triggering any failure during debugging.
```

Comment:
I'd remove "official" here. When you say "official documentation" I'd think that's the upstream docs of the tool itself, not our docs in the scipy repo.